### ISSUE TITLE
Sort linearized match ops to maximize trie prefix sharing

### ISSUE THREAD

**Issue by fitzgen** [cfallin/isle#11]
After linearizing rule patterns for a given term, but before we insert those linear match ops into the trie, we should do a custom topological sort of the linear match ops's data-flow graph that seeks to maximize prefix sharing in the eventual trie:

* create a data-flow graph for each pattern
* keep track of "emittable" match operations for each pattern, i.e. those operations whose operands are all already emitted so we can emit this operation if we choose
* count which operations are most emittable across all patterns
* select the operation with the highest count
* for all patterns where this is an emittable operation:
  * emit this operation next
  * recurse
* for all patterns where that is not an emittable operation:
  * choose the operation with the highest count among these remaining patterns
  * emit this operation next for these patterns
  * recurse
* repeat previous until all patterns have been accounted for

This is greedy algorithm is not optimal, because we don't break ties between equally-emittable operations such that the final trie will have maximal prefix sharing. I can't think of an efficient way to do that right now. But I think this should get us very close to optimal in practice.

**Comment by cfallin** [comment]
I agree, this would be a great improvement to our codegen! The basic idea of emitting all patterns concurrently and scheduling ops based on the ones that have the most "demand" seems sound and pretty robust to me.

Re: this:

> This is greedy algorithm is not optimal, because we don't break ties between equally-emittable operations such that the final trie will have maximal prefix sharing. I can't think of an efficient way to do that right now.

I think that greedy might actually be the best we can do, because this problem in full generality is I think NP-hard; at least, my intuition is telling me that one could reduce a sort of ILP or binpacking problem to this, basically because there's no optimal substructure property; each decision to commit to an op at a point in the tree leads to many possible further subsplits, some of which are more or less expensive than others and these ranges aren't disjoint across the two sides of the decision, so you need to backtrack to solve optimally. Anyway, that feels like an SMT solver pursuing a function-maximization goal...

... all of that to say, I think the above will be more than good enough for what we need :-)
